modules/core/src/test/java/org/apache/ignite/internal/processors/cache/transactions/TxSavepointsTransactionalCacheTest.java

```diff
@@ -0,0 +1,562 @@
+                        " " + isolation + " transaction.", (Integer) 1, cache.get(1));
+
+                    tx.commit();
+                } finally {
+                    cache.remove(1);
+                }
+            }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testPut() throws Exception {
+        final IgniteCache<Integer, Integer> cache = grid(2).getOrCreateCache(getConfig());
+
+        GridTestSafeThreadFactory factory = new GridTestSafeThreadFactory(cacheMode().name()+"_put");
+
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                try (Transaction tx = grid(2).transactions().txStart(concurrency, isolation)) {
+                    info(concurrency + " " + isolation);
+
+                    tx.savepoint("sp");
+
+                    assertTrue(cache.putIfAbsent(1, 0));
+
+                    Thread t = factory.newThread(new Runnable() {
+                        @Override public void run() {
+                            assertTrue(cache.putIfAbsent(1, 1));
+                        }
+                    });
+
+                    t.start();
+
+                    //We need to make sure that second thread trying to get lock.
+                    if (concurrency == TransactionConcurrency.PESSIMISTIC)
+                        while (t.getState() != Thread.State.WAITING)
+                            factory.checkError();
+
+                    tx.rollbackToSavepoint("sp");
+
+                    t.join(3_000);
+
+                    assertEquals("Broken multithreaded rollback to savepoint in " + concurrency +
+                        " " + isolation + " transaction.", (Integer) 1, cache.get(1));
+
+                    tx.commit();
+
+                    assertEquals("Broken rollback to savepoint in " + concurrency + " " + isolation +
+                        " transaction.", (Integer) 1, cache.get(1));
+                } finally {
+                    cache.remove(1);
+                }
+            }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testRemove() throws Exception {
+        final IgniteCache<Integer, Integer> cache = grid(2).getOrCreateCache(getConfig());
+
+        GridTestSafeThreadFactory factory = new GridTestSafeThreadFactory(cacheMode().name()+"_remove");
+
+        for (TransactionConcurrency concurrency : TransactionConcurrency.values())
+            for (TransactionIsolation isolation : TransactionIsolation.values()) {
+                cache.put(1, 1);
+
+                try (Transaction tx = grid(2).transactions().txStart(concurrency, isolation)) {
+                    info(concurrency + " " + isolation);
+
+                    tx.savepoint("sp");
+
+                    assertTrue(cache.remove(1));
+
+                    Thread t = factory.newThread(new Runnable() {
```

Comment:
you should use org.apache.ignite.testframework.GridTestUtils#runMultiThreadedAsync(java.util.concurrent.Callable<?>, int, java.lang.String) instead